src/h5web/providers/jupyter/api.ts

```diff
@@ -0,0 +1,155 @@
+import { Group, Dataset, Metadata, EntityKind, Link } from '../models';
+import type { ProviderAPI } from '../context';
+import {
+  assertGroupContent,
+  isDatasetResponse,
+  isGroupResponse,
+} from './utils';
+import type {
+  JupyterAttrsResponse,
+  JupyterContentResponse,
+  JupyterDataResponse,
+  JupyterMetaResponse,
+} from './models';
+import { nanoid } from 'nanoid';
+import { makeStrAttr, floatType } from '../mock/utils';
+import {
+  HDF5Id,
+  HDF5LinkClass,
+  HDF5ShapeClass,
+  HDF5Value,
+  HDF5SoftLink,
+} from '../hdf5-models';
+import { assertGroup } from '../../guards';
+
+export class JupyterApi implements ProviderAPI {
+  public readonly domain: string;
+  private readonly client: AxiosInstance;
+
+  private values: Record<HDF5Id, HDF5Value> = {};
+
+  public constructor(url: string, domain: string) {
+    this.domain = domain;
+    this.client = axios.create({
+      baseURL: `${url}/hdf`,
+    });
+  }
+
+  public async getMetadata(): Promise<Metadata> {
+    const rootId = '/';
+    const rootGrp = await this.processEntity(rootId);
+    assertGroup(rootGrp);
+    return rootGrp;
+  }
+
+  public async getValue(id: HDF5Id): Promise<HDF5Value | undefined> {
+    if (id in this.values) {
+      return this.values[id];
+    }
+
+    const value = await this.fetchValue(id);
+    this.values[id] = value;
+    return this.values[id];
+  }
+
+  private async fetchAttributes(path: string): Promise<JupyterAttrsResponse> {
+    const { data } = await this.client.get<JupyterAttrsResponse>(
+      `/attrs/${this.domain}?uri=${path}`
+    );
+    return data;
+  }
+
+  private async fetchMetadata(path: string): Promise<JupyterMetaResponse> {
+    const { data } = await this.client.get<JupyterMetaResponse>(
+      `/meta/${this.domain}?uri=${path}`
+    );
+    return data;
+  }
+
+  private async fetchContents(path: string): Promise<JupyterContentResponse> {
+    const { data } = await this.client.get<JupyterContentResponse>(
+      `/contents/${this.domain}?uri=${path}`
+    );
+    return data;
+  }
+
+  private async fetchValue(path: string): Promise<HDF5Value> {
+    const { data } = await this.client.get<JupyterDataResponse>(
+      `/data/${this.domain}?uri=${path}`
+    );
+    return data;
+  }
+
+  /** The main tree-building method */
+  private async processEntity(
+    path: string
+  ): Promise<Group | Dataset | Link<HDF5SoftLink>> {
+    const response = await this.fetchMetadata(path);
+    const { attributeCount } = response;
+
+    const attrReponse =
+      attributeCount > 0 ? await this.fetchAttributes(path) : {};
+    const attributes = Object.entries(attrReponse).map(([name, value]) =>
+      // TODO: fix this once I can infer the type of attributes
+      makeStrAttr(name, value)
+    );
+
+    if (isGroupResponse(response)) {
+      const { name, type, childrenCount } = response;
+      const contents = childrenCount > 0 ? await this.fetchContents(path) : [];
+      assertGroupContent(contents);
```

Comment:
I reckon you could rename `fetchContents` to `fetchGroupContents` and assume that you get a group content response. We already know we're fetching a group and we won't need it for other entities.

Reply:
If you want to keep the exact mapping between the API and the `fetch` methods, you could instead use a generic: `this.fetchContents<JupyterContentGroupResponse>()`

Reply:
I know that it is double-checking but I like the explicit `assert`.

`fetchGroupContents` despite its name could be misused. The `assert` has the merit of making things really clear and without ambiguity.

For the generic, from what I understand, I cannot infer the type from the parameter which leads to a disguised type assertion: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes